### PR TITLE
Rust implementation of svdtools

### PR DESCRIPTION
**Pull request by burrbull**

### CODE REVIEW

**Comment by burrbull** [comment]
cc @adamgreig  @newAM 
Need help with CI.

**Review by newAM — CHANGES_REQUESTED** [review]
I am not a maintainer for this repo, but I am happy to help out with CI anyway.  This is what I would do for CI:

```yaml
on:
  push:
    branches:
      - master
    tags:
      - '*'
  pull_request:

name: Rust CI

jobs:

  build:
    name: Build
    runs-on: ubuntu-latest
    env:
      RUSTFLAGS: "-D warnings"
      RUSTDOCFLAGS: "-D warnings"
    steps:
      - uses: actions/checkout@v2
      - uses: actions-rs/toolchain@v1
        with:
          toolchain: stable
      - run: cargo build
      - run: cargo doc

  test:
    name: Test
    env:
      RUSTFLAGS: "-D warnings"
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions-rs/toolchain@v1
        with:
          toolchain: stable
      - run: cargo test

  clippy:
    name: Clippy
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions-rs/toolchain@v1
        with:
          toolchain: stable
          components: clippy
      # not using clippy-check due to this issue:
      # https://github.com/actions-rs/clippy-check/issues/2
      - run: cargo clippy -- --deny warnings

  format:
    name: Format
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions-rs/toolchain@v1
        with:
          toolchain: stable
          components: rustfmt
      - run: cargo fmt -- --check
```

CI is typically an opinionated topic, so the maintainers may want something different then what I would do.

File: src/patch/mod.rs [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+
+    // Process device
+    svd.process(root, true)
+        .with_context(|| format!("Processing device `{}`", svd.name))?;
+
+    // SVD should now be updated, write it out
+    let svd_out = svd_encoder::encode(&svd)?;
+
+    let mut f = File::create(&svdpath_out)?;
+    f.write(svd_out.as_bytes())?;
```
[the remark this answers is not in the document]
Comment (burrbull) [reply src/patch/mod.rs]: Forgot about clippy. Fixed

**Review by newAM — APPROVED** [review]
Time to run `make patch` on `stm32-rs`:
* python implementation on `stm32-rs`: 7.7s all core / 96.1s single core
* rust implementation on `stm32-rs`: 3.3s all core / 33.8s single core

That is a pretty good performance improvement.

---

There are some warnings, but these look like they should be resolved in stm32-rs:

```text
In String("_modify"): dublicate rule Hash({String("ITR1_RMP"): Hash({String("bitOffset"): Integer(10), String("bitWidth"): Integer(2)})}), ignored
In String("_modify"): dublicate rule Hash({String("ITR1_RMP"): Hash({String("bitOffset"): Integer(10), String("bitWidth"): Integer(2)})}), ignored
In String("_modify"): dublicate rule Hash({String("ITR1_RMP"): Hash({String("bitOffset"): Integer(10), String("bitWidth"): Integer(2)})}), ignored
In String("_modify"): dublicate rule Hash({String("ITR1_RMP"): Hash({String("bitOffset"): Integer(10), String("bitWidth"): Integer(2)})}), ignored
In String("_modify"): dublicate rule Hash({String("ITR1_RMP"): Hash({String("bitOffset"): Integer(10), String("bitWidth"): Integer(2)})}), ignored
```

There are also a couple new errors with the rust version.

```text
error: Processing device `STM32G0B1`

Caused by:
    0: Adding peripheral `DMA1`
    1: device already has a peripheral DMA1
Processing device `STM32G0C1`

Caused by:
    0: Adding peripheral `DMA1`
    1: device already has a peripheral DMA1
```

I did not throughly investigate, I _think_ the python version is just overwriting the previous value whereas the rust version is detecting that this is a problem.
The errors are more desireable than what the python one currently does, so this is fine by me.

---

Overall looks good to me.

The output is identical for the chips that generate without errors.

**Comment by burrbull** [comment]
Added stm32-rs check in Rust CI.

**Review by adamgreig — APPROVED** [review]

**Comment by adamgreig** [comment]
Thanks for all the work on this @burrbull and @MarcoIeni! I think let's get it merged now it's essentially good to go, and we can work on any other improvements in separate PRs.

**Pull request merged**